build-system/tasks/e2e/describes-e2e.js

```diff
@@ -16,19 +16,102 @@
 
 // import to install chromedriver
 require('chromedriver'); // eslint-disable-line no-unused-vars
+
+const puppeteer = require('puppeteer');
 const {AmpDriver, AmpdocEnvironment} = require('./amp-driver');
 const {Builder, Capabilities} = require('selenium-webdriver');
 const {clearLastExpectError, getLastExpectError} = require('./expect');
 const {installRepl, uninstallRepl} = require('./repl');
-const {SeleniumWebDriverController} = require(
-    './selenium-webdriver-controller');
+const {PuppeteerController} = require('./puppeteer-controller');
+const {SeleniumWebDriverController} = require('./selenium-webdriver-controller');
 
 /** Should have something in the name, otherwise nothing is shown. */
 const SUB = ' ';
 const TIMEOUT = 20000;
 
 const DEFAULT_E2E_INITIAL_RECT = {width: 800, height: 600};
 
+/**
+ * @typedef {{
+ *  headless: boolean,
+ *  engine: string,
+ * }}
+ */
+let DescribesConfigDef;
+
+/**
+ * @typedef {{
+ *  headless: boolean,
+ * }}
+ */
+let PuppeteerConfigDef;
+
+/**
+ * @typedef {{
+ *  headless: boolean,
+ * }}
+ */
+let SeleniumConfigDef;
+
+/** @const {!DescribesConfigDef} */
+const describesConfig = {};
+
+/**
+ * Configure all tests.
+ * @param {!DescribesConfigDef} config
+ */
+function configure(config) {
+  Object.assign(describesConfig, config);
```

Comment:
This feels a bit weird, it looks like the config is global. What happens if there are two calls to this configure, with a latter one having only a partial config? For nesting, that can make sense, but what if there are two parallel configs (say two top-level, or two siblings of the top-level), then one would carry over part of the other's config.

Not sure if this is really the case or how it is used, but this pattern looks very suspicious.

Reply:
I implemented it this way for ease of implementation, since its function is mainly to be called once to pass in command line arguments to the `describes` singleton. I agree that long-term this is a code smell. The alternative would be to provide a mechanism to create a `describes` instance in `build-system/tasks/e2e/index.js`, but that seemed like it was a larger change that would be harder to review in combination with this PR

Reply:
Since this isn't safe to call multiple times, I would add a comment (or better an assert) to note that.

Reply:
done